### PR TITLE
Fixing UI tests

### LINKED ISSUE #165: UI: npm test fails with an error
`npm test `

fails with an error:
`
...BaGet.UI\node_modules\office-ui-fabric-react\lib\Icon.js:1
({"Object.<anonymous>":function(module,exports,require,__dirname,__filename,global,jest){export * from './components/Icon/index';
                                                                                         ^^^^^^

SyntaxError: Unexpected token export

  1 | import { HtmlRenderer, Parser } from 'commonmark';
> 2 | import { Icon } from 'office-ui-fabric-react/lib/Icon';
    | ^
  3 | import * as React from 'react';
  4 | import timeago from 'timeago.js';
  5 | import LicenseInfo from './LicenseInfo';

  at ScriptTransformer._transformAndBuildScript (node_modules/jest-runtime/build/script_transformer.js:403:17)
  at Object.<anonymous> (src/DisplayPackage.tsx:2:1)
`

similar to this https://github.com/facebook/jest/issues/2550

I have updated npms to the latest but it does not resolve the issue.

### PR DESCRIPTION
**Pull request by sipmann**
Fixing UI tests and adding new ones.

 * Fixed the test for App component.
 * Coded a better test for App component and SearchResults.
 * Changed the test command to fix the `async` of jest.

Fixes #165 

Could receive more tests. I'm not sure why it's adding the commits I've already pushed to the master branch (eg. improvement of empty lists and 404 pages).

### CODE REVIEW

File: src/BaGet.UI/package.json [hunk 1]
```diff
@@ -25,17 +25,20 @@
...
+    "@testing-library/react": "^10.2.1",
     "@types/commonmark": "^0.27.4",
     "@types/jest": "^23.3.14",
     "@types/node": "^10.17.26",
     "@types/react": "^16.9.36",
     "@types/react-dom": "^16.9.8",
     "@types/react-router-dom": "^5.1.5",
     "@types/semver": "^6.2.1",
+    "jest-environment-jsdom-sixteen": "^1.0.3",
+    "jest-fetch-mock": "^3.0.3",
```
Comment (sipmann) [inline src/BaGet.UI/package.json:41]: Fetch mock more easly

File: src/BaGet.UI/src/App.test.tsx [hunk 2]
```diff
@@ -1,9 +1,33 @@
+// TODO: remove ' --env=jest-environment-jsdom-sixteen' from the npm test script when react-sctips get updated with newer jest version
```
Comment (sipmann) [inline src/BaGet.UI/src/App.test.tsx:1]: Is this the best place to keep track of the workaround?

File: src/BaGet.UI/src/config.tsx [hunk 3]
```diff
@@ -2,6 +2,10 @@ let config = {
   apiUrl: "__BAGET_PLACEHOLDER_API_URL__"
 };
 
+if (process.env.NODE_ENV === 'test') {
```
Comment (sipmann) [inline src/BaGet.UI/src/config.tsx:5]: need this to solve node fetch. It needs a full URL.

File: src/BaGet.UI/src/App.test.tsx (hunk 2 above)
Comment (loic-sharma) [reply src/BaGet.UI/src/App.test.tsx:1]: Yup, I think this works. I'd also create an issue on BaGet and link the URL to the issue here.

File: src/BaGet.UI/src/App.tsx [hunk 4]
```diff
@@ -61,6 +61,7 @@ class App extends React.Component<RouteComponentProps, IAppState> {
                 className="form-control"
                 autoComplete="off"
                 placeholder="Search packages..."
+                data-testid="input-field"
```
Comment (loic-sharma) [inline src/BaGet.UI/src/App.tsx:64]: What do you think of a name like `search-field` to make its purpose a little clearer?

File: src/BaGet.UI/src/SearchResults.test.tsx [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    const { getByText } = render(<Router><SearchResults input='Baget.Unknown' /></Router>);
+
+    const notFoundMessage = await waitFor(
+      () => getByText('Oops, nothing here...')
+    );
+
+    expect(notFoundMessage).toBeDefined();
+  });
+
+  it('Should render a no package found when invalid name was given', async () => {
```
Comment (loic-sharma) [inline src/BaGet.UI/src/SearchResults.test.tsx:33]: It looks like this test name should be updated

File: src/BaGet.UI/src/config.tsx (hunk 3 above)
Comment (loic-sharma) [reply src/BaGet.UI/src/config.tsx:5]: Could you leave a comment about this?

Also, what do you think of only setting `apiUrl` to `http://localhost` if the `apiUrl` currently is `__BAGET_PLACEHOLDER_`?

**Comment by loic-sharma** [comment]
This is awesome!! Thanks for doing this!

> Could receive more tests. 

Small incremental improvements are good! I'd much rather review small chunks than one enormous change 😄 

> I'm not sure why it's adding the commits I've already pushed to the master branch (eg. improvement of empty lists and 404 pages).

I'm not sure either, but the diff looks fine. I'll squash your commits into a single one when I merge this in.

File: src/BaGet.UI/src/App.tsx (hunk 4 above)
Comment (sipmann) [reply src/BaGet.UI/src/App.tsx:64]: perfect xD

**Comment by sipmann** [comment]
I think I've covered every suggestion :D

File: src/BaGet.UI/package.json [hunk 6]
```diff
@@ -49,5 +52,10 @@
       "last 1 firefox version",
       "last 1 safari version"
     ]
+  },
+  "jest": {
+    "moduleNameMapper": {
+      "office-ui-fabric-react/lib/(.*)$": "office-ui-fabric-react/lib-commonjs/$1"
```
Comment (loic-sharma) [inline src/BaGet.UI/package.json:58]: Just for my own knowledge, what does this do?

**Comment by loic-sharma** [comment]
Looks great! It looks like you'll need to press the "Ready for review" button for me to be able to merge this in. The merge button is currently greyed out for me.

File: src/BaGet.UI/package.json (hunk 6 above)
Comment (sipmann) [reply src/BaGet.UI/package.json:58]: The office-ui-fabric uses a different approach to export it's components and the Jest lib isn't able to load it. Take a look at this issue: [5733](https://github.com/microsoft/fluentui/issues/5733)

**Comment by loic-sharma** [comment]
Hey, I forgot that we should run these unit tests as part of the CI build to make sure we don't regress them. Could you update these lines:

https://github.com/loic-sharma/BaGet/blob/1070e9aca0c6eda809a3d56b9519678251d09e33/.azure/pipelines/ci-official.yml#L47-L58

I think adding this step after building the frontend should work:

```yaml
    - task: Npm@1
      displayName: Test frontend
      inputs:
          command: test
          workingDir: src/BaGet.UI
```

See: https://docs.microsoft.com/en-us/azure/devops/pipelines/ecosystems/javascript?view=azure-devops&tabs=code#run-unit-tests

I'll merge these changes in once the CI shows the frontend's test pass. Apologies for the new comment at the last minute!

**Comment by sipmann** [comment]
There we go :). a few commits but it's running now

**Comment by loic-sharma** [comment]
Thanks @sipmann for all the awesome work!

**Pull request merged**